linalg/ginkgo.cpp

```diff
@@ -0,0 +1,510 @@
+   {
+      executor = gko::CudaExecutor::create(0, gko::OmpExecutor::create());
+   }
+   else
+   {
+      mfem::err <<
+                " exec_type needs to be one of the three strings: \"reference\", \"cuda\" or \"omp\" "
+                << std::endl;
+   }
+   using ResidualCriterionFactory = gko::stop::ResidualNormReduction<>;
+   residual_criterion             = ResidualCriterionFactory::build()
+                                    .with_reduction_factor(rel_tol)
+                                    .on(executor);
+
+   combined_factory =
+      gko::stop::Combined::build()
+      .with_criteria(residual_criterion,
+                     gko::stop::Iteration::build()
+                     .with_max_iters(max_iter)
+                     .on(executor))
+      .on(executor);
+}
+
+
+
+void
+GinkgoIterativeSolverBase::initialize_ginkgo_log(gko::matrix::Dense<double>* b)
+{
+   // Add the logger object. See the different masks available in Ginkgo's
+   // documentation
+   convergence_logger = gko::log::Convergence<>::create(
+                           executor, gko::log::Logger::criterion_check_completed_mask);
+   residual_logger = std::make_shared<ResidualLogger<>>(executor,
+                                                        gko::lend(system_matrix),b);
+
+}
+
+
+
+void
+GinkgoIterativeSolverBase::apply(Vector &      solution,
+                                 const Vector &rhs)
+{
+   // some shortcuts.
+   using val_array = gko::Array<double>;
+   using vec       = gko::matrix::Dense<double>;
+
+   MFEM_VERIFY(system_matrix, "System matrix not initialized");
+   MFEM_VERIFY(executor, "executor is not initialized");
+   MFEM_VERIFY(rhs.Size() == solution.Size(),
+               "Mismatching sizes for rhs and solution");
+   // Create the rhs vector in Ginkgo's format.
+   std::vector<double> f(rhs.Size());
+   std::copy(rhs.GetData(), rhs.GetData() + rhs.Size(), f.begin());
+   auto b =
+      vec::create(executor,
+                  gko::dim<2>(rhs.Size(), 1),
+                  val_array::view(executor->get_master(), rhs.Size(), f.data()),
+                  1);
+
+   // Create the solution vector in Ginkgo's format.
+   std::vector<double> u(solution.Size());
+   std::copy(solution.GetData(), solution.GetData() + solution.Size(), u.begin());
+   auto x = vec::create(executor,
+                        gko::dim<2>(solution.Size(), 1),
+                        val_array::view(executor->get_master(),
+                                        solution.Size(),
+                                        u.data()),
+                        1);
+
+   // Create the logger object to log some data from the solvers to confirm
+   // convergence.
+   initialize_ginkgo_log(gko::lend(b));
+
+   MFEM_VERIFY(convergence_logger, "convergence logger not initialized" );
+   if (print_lvl>=2)
+   {
+      MFEM_VERIFY(residual_logger, "residual logger not initialized" );
+      solver_gen->add_logger(residual_logger);
+   }
```

Comment:
This should probably be `if (print_lvl == 1)`

Reply:
Thanks. I was a bit careless. :)